R/functions.R

```diff
@@ -395,224 +382,260 @@ vwparams <- function(vwmodel, name) {
+    constructLabelsCB <- function(targets, probabilities, num_row) {
+        # Initialize output vector
+        tmp <- rep("", num_row)
+        # Iterate cost vectors names
+        for(i in 1:length(targets)) {
+            elem_targets <- eval(get(targets[i], envir=parent.frame(n=1)))
+            elem_probability <- eval(get(probabilities[i], envir=parent.frame(n=1)))
+            # Iterate cost vectors individualy (nrow of data)
+            for (j in 1:num_row) {
+                ifelse(is.na(elem_targets[j]),
+                       "",
+                       tmp[j] <- paste(tmp[j],
+                                       paste(i, elem_targets[j], elem_probability[j], sep = ":")))
+            }
         }
+        trimws(tmp)
     }
     
-    flatten <- function(x) {
-        repeat {
-            if(!any(vapply(x,is.list, logical(1)))) return(x)
-            x <- Reduce(c, x)
+    # namespace load with a yaml file
+    if (typeof(namespaces) == "character" && length(namespaces) == 1 &&
+        grepl("yaml$", namespaces)) {
+        # requireNamespace(yaml)
+        print("Using YAML file for loading the namespaces")
+        if (requireNamespace("yaml", quiet=TRUE, as.character=TRUE)) {
+            namespaces <- yaml::yaml.load_file(namespaces)
+        } else {
+            stop("The 'yaml' package is needed.", .Call=FALSE)
         }
     }
-    temp_params <- params
     
-    # Convert different reductions into string with CL arguments
-    reductions_params <- sapply(names(temp_params$reductions), function(reduction_name) {
-        if (length(temp_params$reductions[[reduction_name]]) == 0) {
-            tmp <- paste0("--", reduction_name)
+    # replace all names to avoid conflicts with VW file format
+    names(data) <- parsingNames(names(data))
+    names(namespaces) <- parsingNames(names(namespaces))
+    for (x in names(namespaces)) namespaces[[x]] <- parsingNames(namespaces[[x]])
+    targets <- parsingNames(targets)
+    if (!is.null(probabilities)) probabilities <- parsingNames(probabilities)
+    if (!is.null(weight)) weight <- parsingNames(weight)
+    if (!is.null(base)) base <- parsingNames(base)
+    if (!is.null(tag)) tag <- parsingNames(tag)
+    
+    # Preparing file
+    if(!append)
+        vw_file <- file(file_path,"w")
+    else
+        vw_file <- file(file_path,"a")
+    
+    # Construct vw format for labels, weights, base and tag
+    formatDataVW <- ""
+    argexpr <- c()
+    names_indices <- c(1:ncol(data))
+    names(names_indices) <- names(data)
+    
+    if(!is.null(targets)) {
+        formatDataVW <- paste0(formatDataVW, "%s")
+        if(length(targets) > 1) {
+            if(!is.null(probabilities)) {
+                if(length(targets) != length(probabilities)) {
+                    stop("targets and probabilities should be of the same length")
+                }
+                temp_labels <- paste0("eval(parse(text = 'constructLabelsCB(targets, probabilities, nrow(data))'))")
```

Comment:
Why is `eval(parse(text= ...))` being used here? This is _never_ a good idea. Can this be modified?

Reply:
The same as before.

General idea was that we can use this evaluation inside `data[,spritf()]` call.
And in my opinion this makes converter really flexible in utilizing different label formats.